### PR TITLE
fix(core/overrideRegistry): Handle Override component synchronously, when possible

### PR DESCRIPTION
**Pull request by christopherthielen**
@erikmunson

### CODE REVIEW

File: app/scripts/modules/core/src/overrideRegistry/Overridable.tsx [hunk 1]
```diff
@@ -55,22 +55,28 @@ export function overridableComponent <P extends IOverridableProps, T extends Rea
...
         })
         .map((accountDetails: IAccountDetails) => this.getComponent(accountDetails))
         .takeUntil(this.destroy$)
-        .subscribe((Component) => this.setState({ Component }));
-    }
+        .subscribe((Component) => {
+          if (constructing) {
+            this.state = { Component };
+          } else {
+            this.setState({ Component })
```
Comment (jrsquared) [inline app/scripts/modules/core/src/overrideRegistry/Overridable.tsx]: semicolons make javascript better;

File: app/scripts/modules/core/src/account/account.service.ts [hunk 2]
```diff
@@ -1,5 +1,6 @@
 import { chain, intersection, zipObject } from 'lodash';
 import { ILogService, IPromise, IQResolveReject, IQService, module } from 'angular';
+import { Observable } from 'rxjs/Observable';
```
Comment (anotherchrisberry) [inline app/scripts/modules/core/src/account/account.service.ts]: deep import bad

File: app/scripts/modules/core/src/account/account.service.ts [hunk 3]
```diff
@@ -148,10 +153,7 @@ export class AccountService {
   }
 
   public listAllAccounts(provider: string = null, providerVersion: string = null): IPromise<IAccountDetails[]> {
-    return this.API.one('credentials')
-      .useCache()
-      .withParams({ expand: true })
-      .get()
+    return this.$q.when(this.accounts$.toPromise())
```
Comment (jrsquared) [inline app/scripts/modules/core/src/account/account.service.ts]: semicolons make something something. (we really should have a lint rule for this...)

File: app/scripts/modules/core/src/account/account.service.ts [hunk 4]
```diff
@@ -51,6 +52,10 @@ export interface IAccountZone {
 }
 
 export class AccountService {
+  public accounts$: Observable<IAccountDetails[]> = Observable.defer(() => {
+    const promise = this.API.one('credentials').useCache().withParams({ expand: true }).get();
+    return Observable.fromPromise<IAccountDetails[]>(promise);
+  }).publishReplay(1).refCount();
```
Comment (jrsquared) [inline app/scripts/modules/core/src/account/account.service.ts]: fancy

File: app/scripts/modules/core/src/overrideRegistry/Overridable.tsx [hunk 5]
```diff
@@ -55,22 +55,28 @@ export function overridableComponent <P extends IOverridableProps, T extends Rea
       super(props);
 
       const { accountService } = ReactInjector;
+      let constructing = true;
```
Comment (jrsquared) [inline app/scripts/modules/core/src/overrideRegistry/Overridable.tsx]: makes me feel super dirty, but i can't think of a better way to do this.

Comment (christopherthielen) [reply app/scripts/modules/core/src/overrideRegistry/Overridable.tsx]: yeah, think of a better way plz

Comment (erikmunson) [reply app/scripts/modules/core/src/overrideRegistry/Overridable.tsx]: yeah chris and i have been discussing how gross this is, haven't come up with a better way yet.

on that note — can you drop a comment somewhere in here covering:

1) this is Bad(TM), and we should Do Better Next Time(TM)
2) why we couldn't Do Better Next Time, This Time(TM)

File: app/scripts/modules/core/src/account/account.service.ts (hunk 3 above)
Comment (erikmunson) [reply app/scripts/modules/core/src/account/account.service.ts]: they sure do make something, that's for sure.


wait


is it Java?? does it make Java??

Comment (jrsquared) [reply app/scripts/modules/core/src/account/account.service.ts]: haha, damn split view diff confusing me. jokes on me.

**Comment by erikmunson** [comment]
👍

**Pull request merged**